### PR TITLE
Bigtable: improve list tables spooler

### PR DESCRIPTION
**Pull request by igorbernstein2**
Avoid blocking the event loop. Previously the first page would be fetched
asynchronously, but all of the other pages would be fetched synchronously
which would block the grpc event loop. The new implementation uses future
chaining.

### CODE REVIEW

**Comment by igorbernstein2** [comment]
@kevinsi4508, cna you take a look when you have a chance?

File: google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java [hunk 1]
```diff
@@ -327,18 +329,68 @@ public Table getTable(String tableId) {
...
+          public ListTablesPage apply(ListTablesPagedResponse response) {
+            return response.getPage();
           }
         },
-        MoreExecutors.directExecutor());
+        MoreExecutors.directExecutor()
+    );
+
+    // Fetch the rest of the pages by chaining the futures.
+    ApiFuture<List<com.google.bigtable.admin.v2.Table>> allProtos = ApiFutures
```
Comment (kevinsi4508) [inline google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java]: nit: "ApiFuture<List<com.google.bigtable.admin.v2.Table>> allTables"?

File: google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java [hunk 2]
```diff
@@ -327,18 +329,68 @@ public Table getTable(String tableId) {
...
+            },
+            MoreExecutors.directExecutor()
+        );
+
+    // Wrap all of the accumulated protos.
+    return ApiFutures.transform(allProtos,
+        new ApiFunction<List<com.google.bigtable.admin.v2.Table>, List<TableName>>() {
+          @Override
+          public List<TableName> apply(List<com.google.bigtable.admin.v2.Table> protos) {
+            List<TableName> results = Lists.newArrayListWithCapacity(protos.size());
```
Comment (kevinsi4508) [inline google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java]: nit: I prefer more specific naming than generic ones.

```
for (com.google.bigtable.admin.v2.Table table : tables) {
  names.add(TableName.parse(table.getName()));
}
```

Up to you.

File: google-cloud-clients/google-cloud-bigtable-admin/src/test/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClientTest.java [hunk 3]
```diff
@@ -327,34 +328,52 @@ public void testGetTableAsync() throws Exception {
   @Test
   public void testListTables() {
     // Setup
-    ListTablesRequest expectedRequest = ListTablesRequest.newBuilder()
-        .setParent(INSTANCE_NAME.toString())
-        .build();
-
-    ListTablesPagedResponse expectedResponseWrapper = Mockito.mock(ListTablesPagedResponse.class);
+    com.google.bigtable.admin.v2.ListTablesRequest expectedRequest =
```
[the remark this answers is not in the document]
Comment (kevinsi4508) [reply google-cloud-clients/google-cloud-bigtable-admin/src/test/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClientTest.java]: Do we need fully qualified name?

[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable-admin/src/test/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClientTest.java]: I'm trying to keep all of the proto classes fully qualified, but didn't want to create noise in this PR converting everything. There will be a follow up PR to make this consistent

File: google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java (hunk 1 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java]: I think keeping the unwrapped objects named protos makes it easier to follow

File: google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java (hunk 2 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java]: see above

**Comment by igorbernstein2** [comment]
@pongad this should be ready for you

File: google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java [hunk 4]
```diff
@@ -327,18 +329,68 @@ public Table getTable(String tableId) {
...
-        MoreExecutors.directExecutor());
+        MoreExecutors.directExecutor()
+    );
+
+    // Fetch the rest of the pages by chaining the futures.
+    ApiFuture<List<com.google.bigtable.admin.v2.Table>> allProtos = ApiFutures
+        .transformAsync(
+            firstPageFuture,
+            new ApiAsyncFunction<ListTablesPage, List<com.google.bigtable.admin.v2.Table>>() {
+              List<com.google.bigtable.admin.v2.Table> responseAccumulator = Lists
```
Comment (pongad) [inline google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java]: This is quite formatted properly. Maybe try `google-java-format-diff.py`?

**Pull request merged**